### PR TITLE
feat(cli): The `meltano run` options are now grouped by categories (state, installation, etc.)

### LINKED ISSUE #8750: feature: Group CLI options into categories
### Feature scope

CLI (options, error messages, logging, etc.)

### Description

For example these are the options for `meltano run` as of v3.5.1:

```
Options:
  --dry-run               Do not run, just parse the invocation, validate it,
                          and explain what would be executed.
  --full-refresh          Perform a full refresh (ignore state left behind by
                          any previous runs). Applies to all pipelines.
  --refresh-catalog       Invalidates catalog cache and forces running
                          discovery before this run.
  --no-state-update       Run without state saving. Applies to all pipelines.
  -f, --force             Force a new run even if a pipeline with the same
                          State ID is already present. Applies to all
                          pipelines.
  --state-id-suffix TEXT  Define a custom suffix to autogenerate state IDs
                          with.
  --merge-state           Merges state with that of previous runs.
  --run-id UUID           Use a custom run ID.
  --install               Install the subject plugin(s) automatically.
  --no-install            Do not install the subject plugin(s) automatically.
  --only-install          Only install the subject plugin(s).
  --database-uri TEXT     System database URI.
  --help                  Show this message and exit.
```

It would be nice if we displayed something like

```
Run options:
  --dry-run               Do not run, just parse the invocation, validate it,
                          and explain what would be executed.
  --full-refresh          Perform a full refresh (ignore state left behind by
                          any previous runs). Applies to all pipelines.
  --refresh-catalog       Invalidates catalog cache and forces running
                          discovery before this run.
  --run-id UUID           Use a custom run ID.

State options:
  --no-state-update       Run without state saving. Applies to all pipelines.
  -f, --force             Force a new run even if a pipeline with the same
                          State ID is already present. Applies to all
                          pipelines.
  --state-id-suffix TEXT  Define a custom suffix to autogenerate state IDs
                          with.
  --merge-state           Merges state with that of previous runs.

Installation options:
  --install               Install the subject plugin(s) automatically.
  --no-install            Do not install the subject plugin(s) automatically.
  --only-install          Only install the subject plugin(s).

Global options:
  --database-uri TEXT     System database URI.
  --help                  Show this message and exit.
```

### PR DESCRIPTION
**Pull request by AvinashYerra**
## Description

This PR improves the readability of `meltano run --help` by grouping CLI options into logical categories instead of displaying them in a single flat list.

The options are now organized into:

- **Run options** – options that affect execution behavior (e.g., `--dry-run`, `--timeout`)
- **State options** – options related to state handling (e.g., `--state-strategy`, `--no-state-update`)
- **Installation options** – plugin installation controls (e.g., `--install`, `--no-install`)
- **Global options** – remaining general-purpose options (e.g., `--database-uri`, `--help`)

### Key Changes

- Overrode `format_options` in `PartialInstrumentedCmd` to support grouped help output
- Categorized options based on their functional responsibility
- Moved:
  - `--state-strategy` under **State options**
  - `--timeout` under **Run options**
 
No changes were made to CLI behavior — this only affects help text rendering.

---

## Checklist

- [X] I have read the [contribution guide](https://docs.meltano.com/contribute/merge/)

### Was generative AI tooling used to co-author this PR?

- [X] Yes Generated by ChatGPT following the agentic coding guidelines

## Related Issues

* Closes #8750

## Summary by Sourcery

Group `meltano run` CLI help options into logical sections without changing command behavior.

Enhancements:
- Introduce grouped help output for the `run` command, organizing options into Run, State, Installation, and Global categories for improved readability.

Tests:
- Add CLI tests to verify grouped help headings, option placement within each group, group ordering, absence of empty/extra groups, and allowed options within the Global group.

### CODE REVIEW

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
Tests?

**Comment by AvinashYerra** [comment]
Hi @edgarrmondragon , I couldn't find the issue here. let me know if any changes are required from my end. happy to help

File: src/meltano/cli/utils.py [hunk 1]
```diff
@@ -649,6 +649,55 @@ def invoke(self, ctx) -> None:  # noqa: ANN001
...
+                groups["Installation options"].append(record)
+
+            else:
+                groups["Global options"].append(record)
+
+        for group_name, records in groups.items():
+            if records:
+                formatter.write_heading(group_name)
+                formatter.write_dl(records)
+                formatter.write_paragraph()
```
Comment (edgarrmondragon) [inline src/meltano/cli/utils.py]: nit: is it possible for `records` to be empty? If not, we could skip the conditional.

File: src/meltano/cli/utils.py [hunk 2]
```diff
@@ -649,6 +649,55 @@ def invoke(self, ctx) -> None:  # noqa: ANN001
             ctx.obj["tracker"].track_command_event(CliEvent.started)
         super().invoke(ctx)
 
+    def format_options(self, ctx, formatter):
```
Comment (edgarrmondragon) [inline src/meltano/cli/utils.py]: nit: I wonder how scalable this approach is. Does it mean every addition or removal of options in `meltano.cli.run` would also require changes to this method?

Comment (edgarrmondragon) [inline src/meltano/cli/utils.py]: Let's add type annotations for `ctx` and `formatter`. Let's also add a one-line docstring and decorate the method with `@override` (similar to https://github.com/meltano/meltano/blob/2f6278cc6f1bc99b4786e859401f3c585e129bc3/src/meltano/cli/_didyoumean.py#L22-L27).

**Review by edgarrmondragon — COMMENTED** [review]
Thanks @AvinashYerra, just small nits. Let's also try make Ruff happy.

File: src/meltano/cli/utils.py (hunk 1 above)
Comment (AvinashYerra) [reply src/meltano/cli/utils.py]: Yes, i believe there can be changes/removal of the options available. so keeping the check avoids rendering empty sections.

File: src/meltano/cli/utils.py (hunk 2 above)
Comment (AvinashYerra) [reply src/meltano/cli/utils.py]: By default, any new options would fall under the "Global options" category. If needed, they can be explicitly grouped by updating this method.
Happy to refactor this to a more scalable approach if you have something specific in mind.

Comment (edgarrmondragon) [reply src/meltano/cli/utils.py]: Would the tests catch this? i.e. another contributor adds a new option to `meltano run` but they forgot to _categorize_ them here.

Comment (AvinashYerra) [reply src/meltano/cli/utils.py]: Done.

Comment (AvinashYerra) [reply src/meltano/cli/utils.py]: Before my latest commit : No tests wouldn't catch this, it just checks whether all the options are going into correct group
After my latest commit : Added a new test, which checks only the mentioned run options go into Global Options.
If there is a new option and not categorised into a group the test will be able catch now

Comment (edgarrmondragon) [reply src/meltano/cli/utils.py]: Thanks!

**Review by edgarrmondragon — APPROVED** [review]

File: src/meltano/cli/utils.py [hunk 3]
```diff
@@ -649,6 +650,52 @@ def invoke(self, ctx) -> None:  # noqa: ANN001
...
+            record = param.get_help_record(ctx)
+            if not record:
+                continue
+
+            groups[opt_to_group.get(param.name, "Global options")].append(record)
+
+        for group_name, records in groups.items():
+            if records:  # pragma: no branch
+                with formatter.section(group_name):
+                    formatter.write_dl(records)
```
Comment (edgarrmondragon) [inline src/meltano/cli/utils.py:697]: TIL this context manager does what we want and then some.

**Review by edgarrmondragon — COMMENTED** [review]
Thanks @AvinashYerra!

**Comment by edgarrmondragon** [comment]
@all-contributors please add @AvinashYerra for code

**Comment by edgarrmondragon** [comment]
@sourcery-ai review

**Comment by AvinashYerra** [comment]
Thanks @edgarrmondragon, appreciate your patience and support. Hoping for more collaborations

File: tests/meltano/cli/test_cli_group.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+def test_no_unexpected_options_in_global_group(runner):
+    """Ensure only expected options fall under Global options."""
+    result = runner.invoke(cli, ["run", "--help"])
+    assert result.exit_code == 0, result.output
+
+    output = result.output
+
+    global_idx = output.find("Global options:")
+    global_section = output[global_idx:]
+
```
[the remark this answers is not in the document]
Comment (edgarrmondragon) [reply tests/meltano/cli/test_cli_group.py:116]: The brittleness is intentional
